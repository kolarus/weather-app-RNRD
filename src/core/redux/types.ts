import {Weather} from 'src/shared/api/weather/types';

export interface RootState {
  user: {
    isAuthorized: boolean;
    ui: {
      isAuthorizationInProgress: boolean;
      hasInvalidLoginAttempt: boolean;
    };
    settings: {
      units: string;
      showWeatherFor: number;
    };
  };
  weather: {
    ui: {
      isFetching: boolean;
    };
    weatherData: Nullable<Weather>;
    selectedCity: string;
    selectedCountry: string;
    lastUpdated: string;
  };
  cities: {
    ui: {
      isFetching: boolean;
    };
    cities: Array<Weather>;
  };
}

export interface Action {
  type: string;
  payload: any;
}
