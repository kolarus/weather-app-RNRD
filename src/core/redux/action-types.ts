export const USER_ACTIONS = {
  SET_HAS_INVALID_LOGIN_ATTEMPT: 'SET_HAS_INVALID_LOGIN_ATTEMPT',
  SET_IS_AUTHORIZATION_IN_PROGRESS: 'SET_IS_AUTHORIZATION_IN_PROGRESS',
  SET_IS_AUTHORIZED: 'SET_IS_AUTHORIZED',
  REQUEST_SIGN_IN: 'REQUEST_SIGN_IN',
};
