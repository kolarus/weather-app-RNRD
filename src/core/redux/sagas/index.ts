import {all} from 'redux-saga/effects';

import watchAuthorizeUser from './watch-authorize-user';
import watchFetchWeather from './watch-fetch-weather';
import watchFetchCities from './watch-fetch-cities';

export default function* rootSaga() {
  yield all([watchAuthorizeUser(), watchFetchWeather(), watchFetchCities()]);
}
