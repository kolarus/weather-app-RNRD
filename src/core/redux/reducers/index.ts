import {combineReducers} from 'redux';

import user from './user';
import weather from './weather';
import cities from './cities';

export default combineReducers({user, weather, cities});
