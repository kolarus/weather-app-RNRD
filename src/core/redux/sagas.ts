import {Alert} from 'react-native';
import {put, takeLatest, all} from 'redux-saga/effects';
import signIn from 'src/shared/api/user/sign-in';

import {
  setIsUserAuthorized,
  setIsAuthorizationInProgress,
  setHasInvalidLoginAttempt,
} from './actions/user';
import {USER_ACTIONS} from './action-types';
import {RequestSignInAction} from './actions/user/types';

export function* authorizeUser(action: RequestSignInAction) {
  try {
    yield put(setIsAuthorizationInProgress(true));
    yield signIn(action.payload.login, action.payload.password);
    yield put(setIsAuthorizationInProgress(false));
    yield put(setIsUserAuthorized(true));
  } catch (error) {
    yield put(setIsAuthorizationInProgress(false));
    yield put(setHasInvalidLoginAttempt(true));
    Alert.alert('Login Error', 'Wrong credentials', [{text: 'OK', style: 'cancel'}]);
  }
}

export function* watchAuthorizeUser() {
  yield takeLatest(USER_ACTIONS.REQUEST_SIGN_IN, authorizeUser);
}

export default function* rootSaga() {
  yield all([watchAuthorizeUser()]);
}
