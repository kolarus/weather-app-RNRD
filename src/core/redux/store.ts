import {createStore, applyMiddleware, compose} from 'redux';
import createSagaMiddleware from 'redux-saga';
import {persistStore, persistReducer} from 'redux-persist';
import AsyncStorage from '@react-native-async-storage/async-storage';

import rootSaga from './sagas';
import rootReducer from './reducers';

const composeEnhancers = window.__REDUX_DEVTOOLS_EXTENSION_COMPOSE__ || compose;
const sagaMiddleware = createSagaMiddleware();

const middlewares = [sagaMiddleware];

if (process.env.NODE_ENV === 'development') {
  const {logger} = require('redux-logger');

  middlewares.push(logger);
}

const persistedReducer = persistReducer(
  {
    key: 'root',
    storage: AsyncStorage,
  },
  rootReducer,
);

export const store = createStore(
  persistedReducer,
  composeEnhancers(applyMiddleware(...middlewares)),
);

// @ts-ignore https://github.com/rt2zz/redux-persist/issues/1140
export const persistor = persistStore(store);

sagaMiddleware.run(rootSaga);
