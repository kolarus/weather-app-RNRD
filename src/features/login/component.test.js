import React from 'react';
import {render} from '@testing-library/react-native';
import {Provider} from 'react-redux';
import {createStore} from 'redux';

import Login from './component';

beforeEach(() => {
  jest.useFakeTimers();
});

const initialState = {
  user: {
    ui: {},
  },
};

test('Login should match snapshot', () => {
  const {toJSON} = render(
    <Provider store={createStore((state) => state, initialState)}>
      <Login />
    </Provider>,
  );

  expect(toJSON()).toMatchSnapshot();
});
