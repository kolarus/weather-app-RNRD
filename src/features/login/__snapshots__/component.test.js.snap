// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Login should match snapshot 1`] = `
<View
  style={
    Object {
      "alignItems": "center",
      "backgroundColor": "#080b24",
      "height": "100%",
      "justifyContent": "space-between",
    }
  }
>
  <View
    style={
      Object {
        "alignItems": "center",
        "flexDirection": "column",
        "height": "60%",
        "justifyContent": "flex-end",
        "transform": Array [
          Object {
            "translateY": 0,
          },
        ],
        "width": "100%",
      }
    }
  >
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Login"
      placeholderTextColor="white"
      rejectResponderTermination={true}
      style={
        Array [
          Object {
            "backgroundColor": "gray",
            "borderRadius": 20,
            "color": "white",
            "fontFamily": "PT Sans Narrow",
            "fontSize": 18,
            "height": 50,
            "marginBottom": 30,
            "paddingLeft": 20,
            "width": "90%",
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
    <TextInput
      allowFontScaling={true}
      autoCapitalize="none"
      onChangeText={[Function]}
      placeholder="Password"
      placeholderTextColor="white"
      rejectResponderTermination={true}
      secureTextEntry={true}
      style={
        Array [
          Object {
            "backgroundColor": "gray",
            "borderRadius": 20,
            "color": "white",
            "fontFamily": "PT Sans Narrow",
            "fontSize": 18,
            "height": 50,
            "marginBottom": 30,
            "paddingLeft": 20,
            "width": "90%",
          },
          undefined,
        ]
      }
      underlineColorAndroid="transparent"
      value=""
    />
  </View>
  <View
    accessible={true}
    focusable={true}
    onClick={[Function]}
    onResponderGrant={[Function]}
    onResponderMove={[Function]}
    onResponderRelease={[Function]}
    onResponderTerminate={[Function]}
    onResponderTerminationRequest={[Function]}
    onStartShouldSetResponder={[Function]}
    style={
      Object {
        "alignItems": "center",
        "backgroundColor": "#ff4200",
        "borderRadius": 20,
        "height": 50,
        "justifyContent": "center",
        "marginVertical": 30,
        "opacity": 1,
        "width": "90%",
      }
    }
  >
    <Text
      style={
        Array [
          Object {
            "color": "white",
            "fontFamily": "PT Sans",
            "fontSize": 15,
          },
          Object {
            "fontSize": 30,
          },
          Object {},
        ]
      }
    >
      LOGIN
    </Text>
  </View>
</View>
`;
